File: app/src/main/java/sk/styk/martin/apkanalyzer/business/service/FeaturesService.java
package sk.styk.martin.apkanalyzer.business.service;

import android.content.pm.FeatureInfo;
import android.content.pm.PackageInfo;
import android.support.annotation.NonNull;

import java.util.ArrayList;
import java.util.List;

import sk.styk.martin.apkanalyzer.model.detail.FeatureData;

/**
 * @author Martin Styk
 * @version 12.10.2017.
 */
public class FeaturesService {

    public List<FeatureData> get(@NonNull PackageInfo packageInfo) {

        FeatureInfo[] featureInfos = packageInfo.reqFeatures;
        if (featureInfos == null) {
            return new ArrayList<>(0);
        }

        List<FeatureData> requestedFeatures = new ArrayList<>(featureInfos.length);

        for (FeatureInfo featureInfo : featureInfos) {
            FeatureData data = new FeatureData();
            String name = featureInfo.name == null ? featureInfo.getGlEsVersion() : featureInfo.name;
            data.setName(name);
            data.setRequired(featureInfo.flags == FeatureInfo.FLAG_REQUIRED);

            requestedFeatures.add(data);
        }

        return requestedFeatures;
    }
}

File: app/src/main/java/sk/styk/martin/apkanalyzer/business/analysis/logic/FeaturesService.kt
package sk.styk.martin.apkanalyzer.business.analysis.logic

import android.content.pm.FeatureInfo
import android.content.pm.PackageInfo
import android.support.annotation.WorkerThread
import sk.styk.martin.apkanalyzer.model.detail.FeatureData
import java.util.*

/**
 * @author Martin Styk
 * @version 12.10.2017.
 */
@WorkerThread
class FeaturesService {

    fun get(packageInfo: PackageInfo): List<FeatureData> {

        val featureInfos = packageInfo.reqFeatures ?: return ArrayList(0)

        return featureInfos.mapTo(ArrayList(featureInfos.size)) {
            FeatureData(
                    name = it.name ?: it.glEsVersion,
                    isRequired = (it.flags and FeatureInfo.FLAG_REQUIRED) > 0
            )
        }
    }
}
